sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -701,6 +775,56 @@ private static <DestinationT> int hashDestination(
         .asInt();
   }
 
+  private static class MaybeDestination<DestinationT> {
+    final DestinationT destination;
+    final boolean isValid;
+
+    MaybeDestination(DestinationT destination, boolean isValid) {
+      this.destination = destination;
+      this.isValid = isValid;
+    }
+  }
+  // Utility method to get the dynamic destination based on a record. Returns a MaybeDestination
+  // because some implementations of dynamic destinations return null, despite this being prohibited
+  // by the interface
+  private MaybeDestination<DestinationT> getDestinationWithErrorHandling(
+      UserT input, MultiOutputReceiver outputReceiver, Coder<UserT> inputCoder) throws Exception {
+    try {
+      return new MaybeDestination<>(getDynamicDestinations().getDestination(input), true);
+    } catch (Exception e) {
+      getBadRecordRouter()
+          .route(
+              outputReceiver, input, inputCoder, e, "Unable to get dynamic destination for record");
+      return new MaybeDestination<>(null, false);
+    }
+  }
+
+  // Utility method to format a record based on the dynamic destination. If the operation fails, and
+  // is output to the bad record router, this returns null
+  private @Nullable OutputT formatRecordWithErrorHandling(
+      UserT input, MultiOutputReceiver outputReceiver, Coder<UserT> inputCoder) throws Exception {
+    try {
+      return getDynamicDestinations().formatRecord(input);
+    } catch (Exception e) {
+      getBadRecordRouter()
+          .route(
+              outputReceiver,
+              input,
+              inputCoder,
+              e,
+              "Unable to format record for Dynamic Destination");
+      return null;
+    }
+  }
+
+  private void addErrorCollection(PCollectionTuple sourceTuple) {
+    getBadRecordErrorHandler()
+        .addErrorCollection(
+            sourceTuple
+                .get(BAD_RECORD_TAG)
+                .setCoder(BadRecord.getCoder(sourceTuple.getPipeline())));
```

Comment:
it can't be entirely generic, as getBadRecordErrorHandler is specific to WriteFiles, but I agree that if I'm repeating this pattern a bunch I should look to simplify it